File: TrolleyCoreTests/BaseTestClass.m
//
//  BaseTestClass.m
//  TrolleyCoreTests
//
//  Created by Harry Wright on 28.09.17.
//  Copyright © 2017 Off-Piste. All rights reserved.
//

#import "BaseTestClass.h"
#import "TrolleyHelpers.h"

@import TrolleyCore;

@implementation BaseTestClass

- (instancetype)init {
    if (self = [super init]) {
        self->_timeout = trl_timeout;
    }
    return self;
}

- (void)waitForExpectations {
    [self waitForExpectationsWithHandler:nil];
}

- (void)waitForExpectationsWithHandler:(XCWaitCompletionHandler)handler {
    [self waitForExpectationsWithTimeout:self.timeout handler:handler];
}

- (void)waitForObserver:(__autoreleasing NSNotificationName)name
            withHandler:(notification_handler)handler {
    [[NSNotificationCenter defaultCenter] addObserverForName:name
                                                      object:nil
                                                       queue:nil
                                                  usingBlock:handler];
}

- (void)setUp {
    [super setUp];
    [Trolley setlogging:trl_isLogging];
}

- (void)tearDown {
    [super tearDown];

    if ([Trolley isShopOpen]) {
        [[Trolley shop] deleteAppWithHandler:nil];
    }
}

@end

File: TrolleyCoreTests/BaseTestClass.swift
//
//  BaseTestClass.swift
//  TrolleyCoreTests
//
//  Created by Harry Wright on 28.09.17.
//  Copyright © 2017 Off-Piste. All rights reserved.
//

import XCTest
import TrolleyCore

extension Bundle {
    static let testBundle: Bundle = Bundle(for: BaseTestClass.self)
}

open class BaseTestClass: XCTestCase {

    final var timeout: TimeInterval = trl_timeout

    final func waitForExpectations(handler: XCWaitCompletionHandler? = nil) {
        self.waitForExpectations(timeout: timeout, handler: handler)
    }

    final func waitForObserver(
        _ name: NSNotification.Name,
        handler: @escaping (Notification) -> Void
        )
    {
        NotificationCenter
            .default
            .addObserver(forName: name, object: nil, queue: .main, using: handler)
    }

    final func defaultException(_ name: String = #function) -> XCTestExpectation {
        return self.expectation(description: name)
    }

    open override func tearDown() {
        super.tearDown()
        Trolley.shop?.deleteApp()
    }


    open override func setUp() {
        super.setUp()
        trl_set_log()
    }

    func describeNoThrow(
        _ description: String,
        file: StaticString = #file,
        line: UInt = #line,
        body: () throws -> Void
        )
    {
        do {
            try body()
        } catch {
            XCTFail(description, file: file, line: line)
        }
    }

    func describeThrow(
        _ description: String,
        file: StaticString = #file,
        line: UInt = #line,
        body: () throws -> Void
        )
    {
        do {
            try body()
            XCTFail(description, file: file, line: line)
        } catch _ { }
    }

}
